### PR TITLE
6969 fix vaults error recovery

### LINKED ISSUE #6969: incomplete recovery from hitting vaults minting limit
## Describe the bug

While working on the vaults UI, @samsiegart hit the minting limit and was surprised that his balance didn't recover. Chain log:
https://gist.github.com/samsiegart/2c1a4a563a2fc96fdd7d4383a062f0fc

## To Reproduce

???

1. ~~check out `1323a207983fbdb69aa09752d613d8d695324d4a` aka `dc-refund-durable `.~~

_that test turned out to be faulty; adding `await eventLoopIteration();` made it pass_

## Expected behavior

funds returned

## Platform Environment

 - what OS are you using? what version of Node.js?
 - is there anything special/unusual about your platform?
 - what version of the Agoric-SDK are you using? (run `git describe --tags --always`)
Sam was running 1323a207983fbdb69aa09752d613d8d695324d4a when he reported the problem.

## Additional context

_IOU pointer to the issue sam was working on_

## Screenshots

_sam shared a few internally..._

### PR DESCRIPTION
**Pull request by turadg**
closes: #6969

## Description

When a vault failed to init, the seat wouldn't be exited. This resulted in the funds hanging on the seat and not being returned to the caller. The fix was to move the `exit()` to `finally`.
- [fix(vaults): exit seat after failure](https://github.com/Agoric/agoric-sdk/pull/7020/commits/2321d8a7ed7667ee0397c9283fc88c31ec13bf8a)

This then sometimes errored with stagings due to https://github.com/Agoric/agoric-sdk/issues/7024 . The solution is to stop using stages, so this does that for vaultFactory as part of https://github.com/Agoric/agoric-sdk/issues/7025. It also does it for stakeFactory because they have a coupling through a shared a dependency on `chargeInterest`.
- [fix: replace staging with atomic transfer](https://github.com/Agoric/agoric-sdk/pull/7020/commits/9db9fadfd1e1391d2ede5607d0c9d89bf2a0da11)

Along the way I tried making a smart wallet integration test to detect the #6969 but it's only with vaults and we can't test that integration until #6880. I left the test in anyway.
- [test(smartWallet): failure retains bank balances](https://github.com/Agoric/agoric-sdk/pull/7020/commits/00c2b037cfad6d8e0c2309da9b36c60daf1b457d)

This also has a CLI feature that was helpful while debugging,
- [feat(wallet): show bank balances](https://github.com/Agoric/agoric-sdk/pull/7020/commits/bcebea9949b38f55f78447824e4bc67c99da6ef6)

And a drive-by for a different state problem,
- [fix(vaults): failures don't consume a vaultId](https://github.com/Agoric/agoric-sdk/pull/7020/commits/ed7899ef50e8d7d01523da8e60ae121bae53f7f6)

### Security Considerations

Prevents a theoretical exploit,
- https://github.com/Agoric/agoric-sdk/issues/7024

### Scaling Considerations

--

### Documentation Considerations

--

### Testing Considerations

Using `start-local-chain` made vaults that used to error and lose funds. Then made the fix and verified the funds are restored after error.

### CODE REVIEW

**Comment by dckc** [comment]
This motivates giving clients the ability to exit #6906.

File: packages/agoric-cli/src/commands/wallet.js [hunk 1]
```diff
@@ -150,9 +151,13 @@ export const makeWalletCommand = async () => {
       try {
         const summary = summarize(current, coalesced, agoricNames);
         process.stdout.write(fmtRecordOfLines(summary));
+        process.stdout.write('\n');
       } catch (e) {
         console.error('CAUGHT HERE', e);
       }
+      execFileSync('agd', [`query`, 'bank', 'balances', opts.from], {
```
Comment (dckc) [inline packages/agoric-cli/src/commands/wallet.js:158]: indexing under ambient authority #2160

(not to mention sync io)

File: packages/inter-protocol/src/vaultFactory/vaultManager.js [hunk 2]
```diff
@@ -863,6 +864,8 @@ export const prepareVaultManagerKit = (
                 vaultId,
                 'never stored during initVaultKit failure',
               );
+            } finally {
```
Comment (dckc) [inline packages/inter-protocol/src/vaultFactory/vaultManager.js]: this seems to be a different try/catch than the one that used to contain `seat.exit()`

**Review by dckc — COMMENTED** [review]
notes so far...

File: packages/inter-protocol/src/vaultFactory/vaultManager.js (hunk 2 above)
Comment (erights) [reply packages/inter-protocol/src/vaultFactory/vaultManager.js]: Good catch! My eye was fooled into doing the wrong curly matching for some reason.

Comment (turadg) [reply packages/inter-protocol/src/vaultFactory/vaultManager.js]: Good eyes! Thankfully the tests caught this too

File: packages/inter-protocol/src/stakeFactory/stakeFactoryKit.js [hunk 3]
```diff
@@ -296,9 +297,17 @@ const helperBehavior = {
...
-    stageDelta(clientSeat, vaultSeat, give, emptyDebt, KW.Debt);
-    manager.mintAndReallocate(toMint, fee, clientSeat, vaultSeat);
+    manager.mintAndReallocate(
+      clientSeat,
+      toMint,
+      fee,
+      harden([
+        [clientSeat, vaultSeat, { [KW.Attestation]: giveColl }],
+        [vaultSeat, clientSeat, { [KW.Attestation]: wantColl }],
+        [clientSeat, vaultSeat, { [KW.Debt]: give }],
```
Comment (Chris-Hibbert) [inline packages/inter-protocol/src/stakeFactory/stakeFactoryKit.js:307]: Could be
```suggestion
        [clientSeat, vaultSeat, { [KW.Attestation]: giveColl, [KW.Debt]: give }],
        [vaultSeat, clientSeat, { [KW.Attestation]: wantColl }],
```
Do you prefer the existing version?

File: packages/inter-protocol/test/vaultFactory/vault-contract-wrapper.js [hunk 4]
```diff
@@ -79,31 +80,41 @@ export async function start(zcf, privateArgs, baggage) {
...
-  const reallocateWithFee = (fee, wanted, seat, ...otherSeats) => {
-    const toMint = AmountMath.add(wanted, fee);
+  /** @type {MintAndTransfer} */
+  const mintAndTransfer = (mintReceiver, toMint, fee, nonMintTransfers) => {
+    const kept = AmountMath.subtract(toMint, fee);
     runMint.mintGains(harden({ Minted: toMint }), stage);
+    /** @type {import('@agoric/zoe/src/contractSupport/atomicTransfer.js').TransferPart[]} */
+    const transfers = [
+      ...nonMintTransfers,
+      [stage, vaultFactorySeat, { Minted: fee }],
```
Comment (Chris-Hibbert) [inline packages/inter-protocol/test/vaultFactory/vault-contract-wrapper.js]: Please rename `stage`, and drop the `stage.clear()` on line 97.  It's no longer being used for staging, and people familiar with stagings will be confused by this naming.

Are there other calls to `.clear()` in vaults or staking that can also be dropped?

File: packages/inter-protocol/test/vaultFactory/vault-contract-wrapper.js [hunk 5]
```diff
@@ -79,31 +80,41 @@ export async function start(zcf, privateArgs, baggage) {
...
-      zcf.reallocate(vaultFactorySeat, stage, seat, ...otherSeats);
+      atomicRearrange(zcf, harden(transfers));
     } catch (e) {
+      console.error('mintAndTransfer caught', e);
       stage.clear();
       vaultFactorySeat.clear();
+      // Make best efforts to burn the newly minted tokens, for hygiene.
+      // That only relies on the internal mint, so it cannot fail without
+      // there being much larger problems. There's no risk of tokens being
+      // stolen here because the staging for them was already cleared.
```
Comment (Chris-Hibbert) [inline packages/inter-protocol/test/vaultFactory/vault-contract-wrapper.js]: There's no staging remaining, right?  What is this referring to?

File: packages/inter-protocol/test/vaultFactory/vault-contract-wrapper.js [hunk 6]
```diff
@@ -79,31 +80,41 @@ export async function start(zcf, privateArgs, baggage) {
...
+      // Note that if this assertion may fail because of an error in the
+      // try{} block, but that error won't be thrown because this executes
+      // before the catch that rethrows it.
       assert(
-        AmountMath.isEmpty(stage.getAmountAllocated('Minted', runBrand)),
-        `Stage should be empty of Minted`,
+        Object.values(stage.getCurrentAllocation()).every(a =>
+          AmountMath.isEmpty(a),
+        ),
+        X`Stage should be empty of Minted`,
```
Comment (Chris-Hibbert) [inline packages/inter-protocol/test/vaultFactory/vault-contract-wrapper.js]: Can we drop this?

**Review by Chris-Hibbert — CHANGES_REQUESTED** [review]

File: packages/inter-protocol/test/vaultFactory/vault-contract-wrapper.js (hunk 5 above)
Comment (turadg) [reply packages/inter-protocol/test/vaultFactory/vault-contract-wrapper.js]: This was just copied over. But that means I should revert the change here and go update vaultDirector and stakeFactory to not talk about staging.

File: packages/inter-protocol/src/stakeFactory/stakeFactoryKit.js (hunk 3 above)
Comment (turadg) [reply packages/inter-protocol/src/stakeFactory/stakeFactoryKit.js:307]: I tried that and the line wrapping made it less readable. I think this style make it more clear what's going on, coming from the stageDelta style.

**Review by Chris-Hibbert — APPROVED** [review]
LGTM.  Very clean. Thanks.

File: packages/zoe/src/contractSupport/atomicTransfer.js [hunk 7]
```diff
@@ -1,8 +1,13 @@
 import { mustMatch, M } from '@agoric/store';
 import { assertRightsConserved } from '../contractFacet/rightsConservation.js';
+import { AmountKeywordRecordShape, SeatShape } from '../typeGuards.js';
 
 const { Fail, quote: q } = assert;
 
+export const TransferPartShape = M.arrayOf(
+  M.or(SeatShape, AmountKeywordRecordShape),
+);
```
Comment (erights) [inline packages/zoe/src/contractSupport/atomicTransfer.js]: Seems surprisingly loose. Trying a tighter pattern at https://github.com/Agoric/agoric-sdk/pull/7040 , but do not yet know what CI has to say about it.

Have not yet looked at the `TransferPart` type, but clearly they should be co-maintained

**Review by erights — CHANGES_REQUESTED** [review]
#7040 green. `TransferPart` type already consistent with suggested `TransferPartShape`

**Comment by erights** [comment]
Only requested changes because I saw this PR was already on the merge train. Now that it's knocked off, I'll approve (with my suggestion), but remove the automerge label first.

**Review by erights — APPROVED** [review]
Not fully reviewed, but approving because I have no objections, and approving is the only way I know to cancel a "Request changes". Because this approval has an outstanding suggestion, I also removed the automerge label first.

**Pull request merged**